cloudsync/cs.py

```diff
@@ -260,24 +260,39 @@ def start(self, *, daemon=True, **kwargs):
         self.nmgr.notify(Notification(SourceEnum.SYNC, NotificationType.STARTED, None))
         self.nmgr.start(**kwargs)
 
-
     def stop(self, forever=True, wait=True):
         """
         Stops the cloudsync service.
 
         Args:
             forever: If false is passed, then handles are left open for a future start.  Generally used for tests only.
         """
+        self._stop(forever=forever, wait=wait, join=True)
+
+    @classmethod
+    def stop_all(cls, syncs: List["CloudSync"]):
+        """
+        Convenience function for stopping multiple CloudSyncs efficiently.
+        """
+        for cs in syncs:
+            cs._stop(forever=True, wait=False, join=False)  # pylint: disable=protected-access
+        for cs in syncs:
+            cs._join()  # pylint: disable=protected-access
+
+    def _stop(self, forever=True, wait=True, join=True):
         log.info("stopping sync: %s", self.storage_label())
         self.smgr.stop(forever=forever, wait=wait)
         self.emgrs[0].stop(forever=forever, wait=wait)
         self.emgrs[1].stop(forever=forever, wait=wait)
         self.nmgr.stop(forever=forever, wait=wait)
+        if join:
+            self._join()
+
+    def _join(self):
         if self.sthread:
             self.sthread.join()
             self.ethreads[0].join()
             self.ethreads[1].join()
-            self.nmgr.wait()  # TODO: wait() above instead of join()?
```

Comment:
Do we not want to join the nmgr thread?

Reply:
I removed this line because it is a noop AFAICT.

The nmgr thread is different than the other 3 in that it is managed by Runnable. The `stop()` call sets Runnable.__thread to `None` so after that there is nothing to wait on.

Reply:
Such BS that it sets the thread to `None`. I want to get rid of that crap so bad